### PR TITLE
Updated help menu keyboard shortcut link to be clickable.

### LINKED ISSUE #2581: help icon last item isn't a link
Moved from https://github.com/CanonicalLtd/jujucharms.com/issues/422

The help icon in the GUI shows three items, the first two are links and the third should be also be a link that opens up the ? keyboard shortcut info. It's odd that two are links and one is not when they all appear the same and there's a hover color change on the links but not the third one.

### LINKED ISSUE #2784: View documentation in help is "useless"
If I land in an empty canvas and click on help/documentation, I'm redirected to getting started docs. Those describe how to install juju and bootstrap a controller. Which is quite useless for the GUI user. 

In case of GISF, we need to redirect them to [1] once it lands in "stable". 

In case of GUI, we need to create better GUI tutorial. 

[1] https://jujucharms.com/docs/devel/getting-started-jaas

### PR DESCRIPTION
**Pull request by Lukewh**
Fixes #2581
Fixes #2784

### CODE REVIEW

**Comment by Lukewh** [comment]
This had a knock on effect that meant some changes to the user menu needed to be made too (around styling).

## QA
- Make sure the 'Keyboard shortcuts' link works in the help menu
- Ensure the user menu still works and looks ok
- Generally ensure nothing is broken ;).

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7119/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7121/

File: jujugui/static/gui/src/app/components/header-help/header-help.js [hunk 1]
```diff
@@ -73,13 +74,44 @@ YUI.add('header-help', function() {
...
+      @param {Object} evt The event that triggered the function
+    */
+    _handleShortcutsLink: function(evt) {
+      const keybindings = this.props.keybindings;
+      // We specifically want the Shift + ? key combo
+      const spec = keybindings['S-/'];
+      const target = document.querySelector(spec.target);
+      if (target) {
+        if (spec.toggle) {
+          if (target.classList.contains('hidden')) {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/header-help/header-help.js]: You could probably just use `target.classList.toggle('hidden');` here.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7140/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7164/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7165/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7166/

File: jujugui/static/gui/src/app/app.js [hunk 2]
```diff
@@ -2904,6 +2849,7 @@ YUI.add('juju-gui', function(Y) {
       this._renderZoom();
       this._renderBreadcrumb();
       this._renderHeaderSearch();
+      this._renderModals();
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: `_renderComponents` gets called every time the db changes (often) do we really need to render the settings modals every time?

File: jujugui/static/gui/src/app/app.js [hunk 3]
```diff
@@ -1254,12 +1185,26 @@ YUI.add('juju-gui', function(Y) {
         document.getElementById('header-search-container'));
     },
 
+    _renderModals: function() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: missing docblock. 
Also maybe something a little less generic `_setupSettingsModals` ?
Then call this at the end of the init vs on every update (see below comment)

Why do we have to have these rendered at all time and then toggle their visibility? As it's a low traffic area I'd image we could get away with rendering/destroying when necessary for both of these modals.

**Comment by Lukewh** [comment]
Added https://github.com/juju/juju-gui/pull/2750 to the fix list

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+YUI.add('modal-gui-settings', function() {
+
+  juju.components.ModalGUISettings = React.createClass({
+
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: plz add `displayName`

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7188/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7189/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7190/

File: jujugui/static/gui/src/app/app.js [hunk 5]
```diff
@@ -1821,6 +1752,44 @@ YUI.add('juju-gui', function(Y) {
...
+          closeModal={this._clearShortcutsModal.bind(this)}
+          keybindings={this.keybindings} />,
+        document.getElementById('modal-shortcuts'));
+    },
+
+    _displaySettingsModal: function() {
+      ReactDOM.render(
+        <window.juju.components.ModalGUISettings
+          closeModal={this._clearSettingsModal.bind(this)}
+          disableCookie={
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: Why are these passed in as well as access to the localStorage? Should the component not just resolve these internally?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7220/

File: jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+
+    @return {Array} An array of the bindings to create ul.
+    */
+    _generateBindings: function() {
+      let bindings = [];
+      Object.keys(this.props.keybindings).forEach(key => {
+        const binding = this.props.keybindings[key];
+        if (binding.help && (binding.condition === undefined ||
+                       binding.condition.call(this) === true)) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js]: odd indentation :)

File: jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+      Generate list of keybindings for help modal
+
+      @return {Array} An array of React li elements.
+    */
+    _generateList: function() {
+      const bindings = this._generateBindings();
+
+      const components = bindings.map((binding) => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js]: extra parens around `binding` and line above

File: jujugui/static/gui/src/app/app.js [hunk 8]
```diff
@@ -108,87 +108,19 @@ YUI.add('juju-gui', function(Y) {
...
-                  }
-                });
-              });
-
-            target.querySelector('.close').addEventListener('click', ev => {
-              document.querySelector('#shortcut-settings').classList.add(
-                'hidden');
-            });
-          }
+        target: '#modal-gui-settings',
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: Are these targets still required now that we're passing the render methods in?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7235/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7239/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7240/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7241/

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+YUI.add('modal-gui-settings', function() {
+
+  juju.components.ModalGUISettings = React.createClass({
+    displayName: 'Custom GUI Settings',
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: this should be the name of the component `ModalGUISettings`

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      Update the state when an input is changed.
+
+      @param {Object} evt the change event.
+    */
+    handleChange: function(evt) {
+      const target = evt.target;
+      const value = target.checked;
+      const name = target.name;
+      this.setState({
+        [name]: value
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: What is this?

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      });
+    },
+
+    /**
+      When the save button is pressed, commit to localStorage.
+    */
+    handleSave: function() {
+      Object.keys(this.state).forEach(function (key) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: arrow fn?

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        if (key !== 'visible') {
+          if (this.state[key] === true) {
+            this.props.localStorage.setItem(key, true);
+          } else {
+            this.props.localStorage.removeItem(key);
+          }
+        }
+      }.bind(this));
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: then this bind can be removed

File: jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+YUI.add('modal-shortcuts', function() {
+
+  juju.components.ModalShortcuts = React.createClass({
+    displayName: 'Keyboard Shortcuts',
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js]: `ModalShortcuts`

File: jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        const binding = this.props.keybindings[key];
+        if (binding.help && (binding.condition === undefined ||
+          binding.condition.call(this) === true)) {
+          bindings.push({
+            key: key,
+            label: binding.label || key,
+            help: binding.help
+          });
+        }
+      }, this);
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/modal-shortcuts/modal-shortcuts.js]: with the arrow fn this `this` isn't required

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js (hunk 10 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: It lets you use a variable as an object key so you don't have to do
```
let state = {};
state[name] = value;
this.setState(state);
```

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7246/

**Comment by Lukewh** [comment]
:shipit: finally

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7247/

**Pull request merged**